cmd/kubeadm/app/cmd/token.go

```diff
@@ -0,0 +1,183 @@
+	createCmd.PersistentFlags().StringVar(
+		&tokenSecret.GivenToken, "token", "",
+		"Shared secret used to secure cluster bootstrap; if none is provided, one will be generated for you",
+	)
+	tokenCmd.AddCommand(createCmd)
+
+	listCmd := &cobra.Command{
+		Use:   "list",
+		Short: "List discovery tokens on the server.",
+		Run: func(tokenCmd *cobra.Command, args []string) {
+			err := RunListTokens(out, tokenCmd)
+			kubeadmutil.CheckErr(err)
+		},
+	}
+	tokenCmd.AddCommand(listCmd)
+
+	deleteCmd := &cobra.Command{
+		Use:   "delete",
+		Short: "Delete discovery tokens on the server.",
+		Run: func(tokenCmd *cobra.Command, args []string) {
+			err := RunDeleteToken(out, tokenCmd, args[0])
+			kubeadmutil.CheckErr(err)
+		},
+	}
+	tokenCmd.AddCommand(deleteCmd)
+
+	return tokenCmd
+}
+
+// RunCreateToken generates a new bootstrap token and stores it as a secret on the server.
+func RunCreateToken(out io.Writer, cmd *cobra.Command, tokenDuration time.Duration, tokenSecret *kubeadmapi.Secrets) error {
+	client, err := kubemaster.CreateClientFromFile(path.Join(kubeadmapi.GlobalEnvParams.KubernetesDir, "admin.conf"))
+	if err != nil {
+		return err
+	}
+
+	err = kubeadmutil.GenerateTokenIfNeeded(tokenSecret)
+	if err != nil {
+		return err
+	}
+
+	secret := &api.Secret{
+		ObjectMeta: api.ObjectMeta{
+			Name: fmt.Sprintf("%s%s", bootstrapTokenSecretPrefix, tokenSecret.TokenID),
+		},
+		Type: api.SecretTypeBootstrapToken,
+		Data: encodeTokenSecretData(tokenSecret, tokenDuration),
+	}
+	if _, err := client.Secrets(api.NamespaceSystem).Create(secret); err != nil {
+		return fmt.Errorf("<cmd/token> failed to bootstrap token [%v]", err)
+	}
+	fmt.Println(tokenSecret.GivenToken)
+
+	return nil
+}
+
+func encodeTokenSecretData(tokenSecret *kubeadmapi.Secrets, duration time.Duration) map[string][]byte {
+	var (
+		data = map[string][]byte{}
+	)
+
+	data["token-id"] = []byte(tokenSecret.TokenID)
+	data["token-secret"] = []byte(tokenSecret.BearerToken)
+
+	t := time.Now()
+	t = t.Add(duration)
+	data["expiration"] = []byte(t.Format(time.RFC3339))
+
+	return data
+}
+
+// RunListTokens lists details on all existing bootstrap tokens on the server.
+func RunListTokens(out io.Writer, cmd *cobra.Command) error {
+	client, err := kubemaster.CreateClientFromFile(path.Join(kubeadmapi.GlobalEnvParams.KubernetesDir, "admin.conf"))
+	if err != nil {
+		return err
+	}
+
+	tokenSelector := fields.SelectorFromSet(
+		map[string]string{
+			api.SecretTypeField: string(api.SecretTypeBootstrapToken),
+		},
+	)
+	listOptions := api.ListOptions{
+		FieldSelector: tokenSelector,
+	}
+
+	results, err := client.Secrets(api.NamespaceSystem).List(listOptions)
+	if err != nil {
+		return fmt.Errorf("<cmd/token> failed to list bootstrap tokens [%v]", err)
+	}
+
+	w := tabwriter.NewWriter(os.Stdout, 10, 4, 3, ' ', 0)
+	fmt.Fprintln(w, "ID\tTOKEN\tEXPIRATION")
+	for _, secret := range results.Items {
+		tokenId := secret.Data["token-id"]
```

Comment:
there is no guarantee that these secrets are well formed.  Use the tuple return form (`tokenId, ok := secret.Data["token-id"]` and check `ok`.

If keys are malformed, I'd output a warning but continue on.  There will be some good keys that folks can use.   That means you probably want to send stderr in too.